text/0000-match-ergonmics.md

```diff
@@ -0,0 +1,531 @@
+
+```
+match e {
+    &E::Foo(...) => { ... }
+    &E::Bar => { ... }
+}
+```
+
+and
+
+```
+match *e {
+    E::Foo(...) => { ... }
+    E::Bar => { ... }
+}
+```
+
+The former is more obvious, but requires more noisey syntax (an `&` on every
+arm). The latter can appear a bit magical to newcomers - the type checker treats
+`*e` as a value, but the borrow checker treats the data as borrowed for the
+duration of the match. It also does not work with nested types, `match (*e,)
+...` for example is not allowed.
+
+In either case if we further bind variables, we must ensure that we do not
+attempt to move data, e.g.,
+
+```
+match *e {
+    E::Foo(x) => { ... }
+    E::Bar => { ... }
+}
+```
+
+If the type of `x` does not have the `Copy` bound, then this will give a borrow
+check error. We must use the `ref` keyword to take a reference: `E::Foo(ref x)`
+(or `&E::Foo(ref x)` if we match `e` rather than `*e`).
+
+The `ref` keyword is a pain for Rust beginners, and a bit of a wart for everyone
+else. It violates the rule of patterns matching declarations, it is not found
+anywhere outside of patterns, and it is often confused with `&`. (See for
+example, https://github.com/rust-lang/rust-by-example/issues/390).
+
+Match expressions are an area where programmers often end up playing 'type
+Tetris': adding operators until the compiler stops complaining, without
+understanding the underlying issues. This serves little benefit - we can make
+match expressions much more ergonomic without sacrificing safety or readability.
+
+Match ergonomics has been highlighted as an area for improvement in 2017:
+[internals thread](https://internals.rust-lang.org/t/roadmap-2017-productivity-learning-curve-and-expressiveness/4097)
+and [Rustconf keynote](https://www.youtube.com/watch?v=pTQxHIzGqFI&list=PLE7tQUdRKcybLShxegjn0xyTTDJeYwEkI&index=1).
+
+
+# Detailed design
+[design]: #detailed-design
+
+I propose two mostly orthogonal improvements: match target auto-deref removes
+the need to write `*e` in the above example; match arm 'auto-ref' infers use
+of `ref` in patterns in match arms, allowing them to be elided. Both features
+aim to make `match` expressions easier to read and write by reducing
+boilerplate.
+
+## Match target auto-deref
+
+The target expression (aka discriminant) in a match expression is `expr` in
+`match expr { ... }`.
+
+If the target expression is borrowed data, then the compiler will implicitly
+dereference the target expression as many times as necessary to make every
+pattern in every arm type check. Dereference here is equivalent to using the `*`
+operator: either a primitive pointer dereference or calling the `deref` method,
+if a type implements the `Deref` trait.
+
+Auto-deref allows the programmer to access borrowed data within a limited scope
+without having to explicitly dereference the data. This is analogous to the
+implicit dereferencing performed by the dot operator, however, the scope of
```

Comment:
What does it mean for the dereference to exist? There is no temporaries that can "stop existing" in both function calls and match expressions. The resulting value lives as long as borrowck lets it live.